packages/xforms-engine/src/parse/xpath/semantic-analysis.ts

```diff
@@ -0,0 +1,290 @@
+	const [, ...argumentNodes] = syntaxNode.children;
+
+	if (argumentNodes.length === 0 && expected.length === 0) {
+		return true;
+	}
+
+	if (argumentNodes.length > expected.length) {
+		return false;
+	}
+
+	return expected.every((types, i) => {
+		const argumentNode = argumentNodes[i];
+
+		if (argumentNode == null) {
+			return false;
+		}
+
+		const [firstMatch] = collectTypedNodes(types, argumentNode);
+
+		return firstMatch != null && isCompleteSubExpression(argumentNode, firstMatch);
+	});
+};
+
+const isTranslationFunctionCall = (syntaxNode: FunctionCallNode): boolean => {
+	return (
+		isCallToLocalNamedFunction(syntaxNode, 'itext') &&
+		hasCallSignature(syntaxNode, [
+			['string_literal', 'absolute_location_path', 'relative_location_path'],
+		])
+	);
+};
+
+export type TranslationExpression = LocalNamedFunctionCallLiteral<'itext'>;
+
+/**
+ * Determines if an arbitrary XPath expression is (in whole) a translation
+ * expression (i.e. a call to `jr:itext`).
+ */
+export const isTranslationExpression = (
+	expression: string
+): expression is TranslationExpression => {
+	const { rootNode } = expressionParser.parse(expression);
+	const functionCallNode = findTypedPrincipalExpressionNode(['function_call'], rootNode);
+
+	if (functionCallNode == null) {
+		return false;
+	}
+
+	return isTranslationFunctionCall(functionCallNode);
+};
+
+const isCurrentFunctionCall = (syntaxNode: FunctionCallNode): boolean => {
+	return isCallToLocalNamedFunction(syntaxNode, 'current') && hasCallSignature(syntaxNode, []);
+};
+
+/**
+ * Predicate to determine if a FilterPathExpr (as currently produced by
+ * `tree-sitter-xpath`) is one of:
+ *
+ * - `current()`
+ * - `current()/...` (where `...` represents additional steps)
+ * - `current()//...` (^)
+ *
+ * @todo XPath grammar technically also allows for `current()[some-predicate]`,
+ * and our `tree-sitter-xpath` grammar/parser also allow for this. But
+ * `@getodk/xpath` types do not currently acknowledge this possibility.
+ */
+export const isCurrentPath = (syntaxNode: FilterPathExprNode): boolean => {
+	const [filterExprNode] = syntaxNode.children;
+	const [anyExprNode] = filterExprNode.children;
+
+	return anyExprNode.type === 'function_call' && isCurrentFunctionCall(anyExprNode);
+};
+
+const isInstanceFunctionCall = (syntaxNode: FunctionCallNode): boolean => {
+	return (
+		isCallToLocalNamedFunction(syntaxNode, 'instance') &&
+		hasCallSignature(syntaxNode, [
+			// Specified as `instance("id")`, but do we really care about the argument's
+			// type here?
+			['argument'],
+		])
+	);
+};
+
+/**
+ * Predicate to determine if a FilterPathExpr (as currently produced by
+ * `tree-sitter-xpath`) is one of:
+ *
+ * - `instance("id")`
+ * - `instance("id")/...` (where `...` represents additional steps)
+ * - `instance("id")//...` (^)
+ *
+ * @todo XPath grammar technically also allows for `current()[some-predicate]`,
```

Comment:
Right. Missed in copypasta. I think it'd be better both this and [the other case](https://github.com/getodk/web-forms/pull/166#discussion_r1684703023)) to reference the grammar names for the syntax, rather than the specific function names. Which may also help clarify some of the questions raised in that other thread.

The point in both cases is that they're not accounting for predicates which may be present, because the current `SyntaxNode` types don't allow for it.